pulsar-jms-filters/src/main/java/com/datastax/oss/pulsar/jms/selectors/JMSPublishFilters.java

```diff
@@ -107,29 +119,63 @@ public void messageProduced(
       long ledgerId,
       long entryId,
       Topic.PublishContext publishContext) {
-    if (!enabled || publishContext.getProperty(JMS_FILTERED_PROPERTY) == null) {
+    if (!enabled) {
+      return;
+    }
+    MessageMetadata messageMetadata =
+        (MessageMetadata) publishContext.getProperty(JMS_FILTER_METADATA);
+    if (messageMetadata == null) {
       return;
     }
-    producer
-        .getTopic()
-        .getSubscriptions()
-        .forEach(
-            (name, subscription) -> {
-              String property = "filter-result-" + name + "@" + subscription.getTopicName();
-              EntryFilter.FilterResult filterResult =
-                  (EntryFilter.FilterResult) publishContext.getProperty(property);
-              if (filterResult == EntryFilter.FilterResult.REJECT) {
-                if (log.isDebugEnabled()) {
-                  log.debug("Reject message {}:{} for subscription {}", ledgerId, entryId, name);
-                }
-                // ir is possible that calling this method in this thread may affect performance
-                // let's keep it simple for now, we can optimize it later
-                subscription.acknowledgeMessage(
-                    Collections.singletonList(new PositionImpl(ledgerId, entryId)),
-                    CommandAck.AckType.Individual,
-                    null);
+    if (messageMetadata.hasNumMessagesInBatch()) {
+      return;
+    }
+    for (Subscription subscription : producer.getTopic().getSubscriptions().values()) {
+      scheduleOnDispatchThread(
+          subscription,
+          () -> {
+            FilterContext filterContext = new FilterContext();
+            filterContext.setSubscription(subscription);
+            filterContext.setMsgMetadata(messageMetadata);
+            filterContext.setConsumer(null);
+            Entry entry = null; // we would need the Entry only in case of batch messages
+            EntryFilter.FilterResult filterResult = filter.filterEntry(entry, filterContext);
```

Comment:
By moving this here, we lose the natural back pressure that could slow down producers, which might lead to overwhelmed filtering. We also lose the ability to persist the result of the filter. I think, but am not certain, that means we have to re-apply the filter on the consumer path because we don't know if the broker failed to run the filter.

Reply:
> that means we have to re-apply the filter on the consumer path because we don't know if the broker failed to run the filter.
we can't skip this, see the description in #123

Reply:
also this thread is the same thread that is currently running filters on the dispatch patch

Reply:
Is the filter header problem the fact that the filter could change? Also, I mean the fact that we changed from `published` to `produced`

Reply:
What do you mean with 'the filter could change' ?
Are you referring to a potential update of the subscription properties? I am not sure it is a big deal.